File: app/src/main/java/com/starostinvlad/fan/LoginScreen/LoginFragment.java
package com.starostinvlad.fan.LoginScreen;


import android.app.ProgressDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.google.android.material.snackbar.Snackbar;
import com.starostinvlad.fan.App;
import com.starostinvlad.fan.R;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public class LoginFragment extends Fragment implements LoginFragmentContract {

    private final String TAG = getClass().getSimpleName();
    private Button regBtn, loginBtn;
    private EditText emailField, passField, nameField;
    private ProgressBar progress;
    private LinearLayout loginContainer;
    private LoginPresenter presenter;
    private boolean isRegistry = false;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);
        loginBtn = view.findViewById(R.id.btn_login);
        regBtn = view.findViewById(R.id.btn_registry);
        emailField = view.findViewById(R.id.email_field);
        passField = view.findViewById(R.id.password_field);
        nameField = view.findViewById(R.id.name_field);
        progress = view.findViewById(R.id.login_progress);
        loginContainer = view.findViewById(R.id.login_container);
        presenter = new LoginPresenter(this);

        ProgressDialog progressDialog = new ProgressDialog(getContext());
        progressDialog.setMessage("Загрузка");

        regBtn.setOnClickListener(v -> {
            openRegistry();
//            isRegistry = !isRegistry;
//            if (isRegistry) {
//                nameField.setVisibility(View.VISIBLE);
//                loginBtn.setText("Регистрация");
//                regBtn.setText("Войти");
//            } else {
//                nameField.setVisibility(View.GONE);
//                regBtn.setText("Регистрация");
//                loginBtn.setText("Войти");
//            }
        });

        nameField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                loginBtn.setEnabled(isValidPass() && isValidEmail() && isValidName());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });


        emailField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (!isRegistry)
                    loginBtn.setEnabled(isValidPass() && isValidEmail());
                else
                    loginBtn.setEnabled(isValidPass() && isValidEmail() && isValidName());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
        passField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (!isRegistry)
                    loginBtn.setEnabled(isValidPass() && isValidEmail());
                else
                    loginBtn.setEnabled(isValidPass() && isValidEmail() && isValidName());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });

        loginBtn.setOnClickListener(view1 -> {
            if (isRegistry)
                presenter.registryApi(emailField.getText().toString(),
                        passField.getText().toString(),
                        nameField.getText().toString()
                );
            else
                presenter.loginApi(emailField.getText().toString(),
                        passField.getText().toString());
        });
        return view;
    }

    private void openRegistry() {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(App.getInstance().getDomain() + "/?do=register"));
        startActivity(browserIntent);
    }

    @Override
    public void showLoading(boolean load) {
        progress.setVisibility(load ? View.VISIBLE : View.GONE);
        loginContainer.setVisibility(!load ? View.VISIBLE : View.GONE);
    }

    private boolean isValidName() {
        String name = nameField.getText().toString();
        return (name.length() > 4 && name.length() < 31);
    }


    private boolean isValidPass() {
        String pass = passField.getText().toString();
        if (!isRegistry)
            return (pass.length() > 5 && pass.length() < 31);
        else
            return (pass.length() > 8 && pass.length() < 31);
    }

    private boolean isValidEmail() {
        String email = emailField.getText().toString();
        return !email.isEmpty();
    }

    @Override
    public void alarm(String message) {
        if (getContext() != null)
            Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/starostinvlad/fan/LoginScreen/LoginFragment.kt
package com.starostinvlad.fan.LoginScreen

import android.app.ProgressDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.starostinvlad.fan.App
import com.starostinvlad.fan.R

class LoginFragment : Fragment(), LoginFragmentContract {
    private val TAG: String = this::class.simpleName!!
    private var regBtn: Button? = null
    private var loginBtn: Button? = null
    private var emailField: EditText? = null
    private var passField: EditText? = null
    private var nameField: EditText? = null
    private var progress: ProgressBar? = null
    private var loginContainer: LinearLayout? = null
    private var presenter: LoginPresenter? = null
    private val isRegistry = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onDestroy() {
        if (presenter != null) {
            presenter!!.detachView()
        }
        super.onDestroy()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_login, container, false)
        loginBtn = view.findViewById(R.id.btn_login)
        regBtn = view.findViewById(R.id.btn_registry)
        emailField = view.findViewById(R.id.email_field)
        passField = view.findViewById(R.id.password_field)
        nameField = view.findViewById(R.id.name_field)
        progress = view.findViewById(R.id.login_progress)
        loginContainer = view.findViewById(R.id.login_container)
        presenter = LoginPresenter()
        presenter!!.attachView(this)
        val progressDialog = ProgressDialog(context)
        progressDialog.setMessage("Загрузка")
        regBtn?.setOnClickListener(View.OnClickListener { v: View? -> openRegistry() })
        nameField?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                loginBtn?.setEnabled(isValidPass && isValidEmail && isValidName)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        emailField?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                if (!isRegistry) loginBtn?.setEnabled(isValidPass && isValidEmail) else loginBtn?.setEnabled(isValidPass && isValidEmail && isValidName)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        passField?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                if (!isRegistry) loginBtn?.setEnabled(isValidPass && isValidEmail) else loginBtn?.setEnabled(isValidPass && isValidEmail && isValidName)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        loginBtn?.setOnClickListener(View.OnClickListener { view1: View? ->
            if (isRegistry) presenter!!.registryApi(emailField?.getText().toString(),
                    passField?.getText().toString(),
                    nameField?.getText().toString()
            ) else presenter!!.loginApi(emailField?.getText().toString(),
                    passField?.getText().toString())
        })
        return view
    }

    private fun openRegistry() {
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("${App.instance.domain}/?do=register"))
        startActivity(browserIntent)
    }

    override fun showLoading(load: Boolean) {
        progress!!.visibility = if (load) View.VISIBLE else View.GONE
        loginContainer!!.visibility = if (!load) View.VISIBLE else View.GONE
    }

    private val isValidName: Boolean
        get() {
            val name = nameField!!.text.toString()
            return name.length > 4 && name.length < 31
        }
    private val isValidPass: Boolean
        get() {
            val pass = passField!!.text.toString()
            return if (!isRegistry) pass.length > 5 && pass.length < 31 else pass.length > 8 && pass.length < 31
        }
    private val isValidEmail: Boolean
        get() {
            val email = emailField!!.text.toString()
            return !email.isEmpty()
        }

    override fun alarm(message: String) {
        if (context != null) Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }
}